### PR TITLE
Allow OpenRAVE to load scenes from remote `http`/`https`/`ftp`/`file` server

### PR DESCRIPTION
**Pull request by ziyan**
Example usage:

```python
env.LoadURI('openrave:/someScene.json', {
    'remoteurl': 'http://username:password@someServer:somePort/someFolder',
})

env.LoadURI('openrave:/someScene.json', {
    'remoteurl': 'file:///some/path/in/local/filesystem',
})
```

Benchmark shows that:

1. loading using `libcurl` from local filesystem using `file:///some/path` is the fastest
2. loading using `ifstream` is second fastest when there is no references
3. loading using `libcurl` from http is faster than `ifstream` when there are references

Screenshot from 2022-10-17 15-07-32

tests in `testopenrave`

### CODE REVIEW

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 1]
```diff
@@ -140,9 +154,12 @@ class JSONReader
     JSONReader(const AttributesList& atts, EnvironmentBasePtr penv, const std::string& defaultSuffix) : _penv(penv), _defaultSuffix(defaultSuffix)
     {
         FOREACHC(itatt, atts) {
-            if (itatt->first == "openravescheme") {
+            if (itatt->first == "openravescheme" || itatt->first == "remotescheme") {
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:157]: @JosephBerman Please remove `remotescheme`

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 2]
```diff
@@ -140,9 +154,12 @@ class JSONReader
...
                 std::stringstream ss(itatt->second);
                 _vOpenRAVESchemeAliases = std::vector<std::string>((istream_iterator<std::string>(ss)), istream_iterator<std::string>());
+                if (itatt->first == "remotescheme") {
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:160]: @JosephBerman In the design, we never added `remotescheme`. Here everything in `_vOpenRAVESchemeAliases` should be remote if `remoteurl` is given.

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 3]
```diff
@@ -155,35 +172,44 @@ class JSONReader
...
         uint64_t starttimeus = utils::GetMonotonicTime();
         vCreatedBodies.clear();
         vModifiedBodies.clear();
         vRemovedBodies.clear();
-        if( !rEnvInfo.IsObject() ) {
+        if( !rEnvInfo.IsObject()) {
             throw OPENRAVE_EXCEPTION_FORMAT("The environment data needs to be a valid dictionary. Currently it is '%s'", orjson::DumpJson(rEnvInfo), ORE_InvalidArguments);
         }
-
+        //Currently reading from environment
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:212]: Comment should have space after `//`, please fix all occurrence.

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 4]
```diff
@@ -192,10 +218,13 @@ class JSONReader
             // extract everything
             _penv->ExtractInfo(envInfo);
         }
-
         std::map<RobotBase::ConnectedBodyInfoPtr, std::string> mapProcessedConnectedBodyUris;
+        std::string referenceUri;
+        if(!rEnvInfo.IsNull()) {
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:223]: Revert this change, `rEnvInfo` is not null, already checked above.

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 5]
```diff
@@ -211,7 +240,7 @@ class JSONReader
                 }
             }
 
-            boost::shared_ptr<const rapidjson::Document> prReferenceEnvInfo = _GetDocumentFromFilename(fullFilename, alloc);
+            boost::shared_ptr<const rapidjson::Document> prReferenceEnvInfo = _GetDocumentFromFilename(fullFilename);
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:243]: You cannot omit `alloc` here, revert the change, you have to pass the `alloc` in.

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 6]
```diff
@@ -237,6 +266,11 @@ class JSONReader
             RAVELOG_ERROR_FORMAT("Failed to load env referenceUri='%s' from file '%s'", referenceUri%_filename);
         }
 
+        //If remote URL is provided, start the process to download everything and load it into the json map
+        if(IsRemote()) {
+            _DownloadRecursively(rEnvInfo);
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:271]: Isn't this too late? `_GetDocumentFromFilename` is already called before this line.

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 7]
```diff
@@ -362,6 +397,38 @@ class JSONReader
         }
     }
 
+    bool IsRemote()
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:400]: `bool IsDownloadingFromRemote() const`

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 8]
```diff
@@ -362,6 +397,38 @@ class JSONReader
...
+
+
+        if (scheme == "file") {
+            return ResolveURI(scheme, path, std::string(), _vOpenRAVESchemeAliases);
+        }
+        else if (_StartsWith(uri, "http")) {
+            return uri;
+        }
+        else if (IsRemote()) {
+            return _remoteUrl + uri.substr(uri.find_first_of(":/")+2,  uri.find_last_of("#") - uri.find_last_of(":/") -1);
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:417]: I do not agree with this design. You should not use `_remoteUrl` until right before calling `curl`.

`_StartsWith(uri, "http")` is a very dangerous check, you are preventing loading of a scene like `http.mujin.msgpack`.

In the future, there could be other protocols, like `remoteUrl = ftp://....` or even `remoteUrl = files://...`, you cannot assume `http`

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 9]
```diff
@@ -384,45 +451,52 @@ class JSONReader
         return false;
     }
 
-    boost::shared_ptr<const rapidjson::Document> _GetDocumentFromFilename(const std::string& fullFilename, rapidjson::Document::AllocatorType& alloc)
+    boost::shared_ptr<const rapidjson::Document> _GetDocumentFromFilename(const std::string& referenceURI)
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:454]: `referenceURI` -> `referenceUri`, remember to follow convention.

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 10]
```diff
@@ -384,45 +451,52 @@ class JSONReader
...
+        if (_rapidJSONDocuments.find(referenceURI) != _rapidJSONDocuments.end()) {
+            doc = _rapidJSONDocuments[referenceURI];
         }
         else {
-            boost::shared_ptr<rapidjson::Document> newDoc;
-            if (_EndsWith(fullFilename, ".json")) {
-                newDoc.reset(new rapidjson::Document(&alloc));
-                OpenRapidJsonDocument(fullFilename, *newDoc);
+            boost::shared_ptr<rapidjson::Document> newDoc = boost::shared_ptr<rapidjson::Document>(new rapidjson::Document());;
+            if (_EndsWith(referenceURI, ".json") && !_StartsWith(referenceURI, "http")) {
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:465]: This is not a good check. Reason already mentioned above.

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 11]
```diff
@@ -384,45 +451,52 @@ class JSONReader
...
+
         return doc;
     }
 
     void _ProcessEnvInfoBodies(EnvironmentBase::EnvironmentBaseInfo& envInfo, const rapidjson::Value& rEnvInfo, rapidjson::Document::AllocatorType& alloc, const std::string& currentUri, const std::string& currentFilename, std::map<RobotBase::ConnectedBodyInfoPtr, std::string>& mapProcessedConnectedBodyUris)
     {
-        dReal fUnitScale = _GetUnitScale(rEnvInfo, 1.0);
+
         std::vector<int> vInputToBodyInfoMapping;
+        dReal fUnitScale = _GetUnitScale(rEnvInfo, 1.0);
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:486]: Revert meaningless changes. They just make your PR harder to review. Revert this

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 12]
```diff
@@ -384,45 +451,52 @@ class JSONReader
...
+            boost::shared_ptr<rapidjson::Document> newDoc = boost::shared_ptr<rapidjson::Document>(new rapidjson::Document());;
+            if (_EndsWith(referenceURI, ".json") && !_StartsWith(referenceURI, "http")) {
+                _urlsAlreadyStaged.push_back(referenceURI);
+                OpenRapidJsonDocument(referenceURI, *newDoc);
             }
-            else if (_EndsWith(fullFilename, ".msgpack")) {
-                newDoc.reset(new rapidjson::Document(&alloc));
-                OpenMsgPackDocument(fullFilename, *newDoc);
+            else if (_EndsWith(referenceURI, ".msgpack")) {
+                OpenMsgPackDocument(referenceURI, *newDoc);
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:470]: By the way, message pack format are loaded into rapidjson too. I think you are confused about it.

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 2 above)
Comment (JosephBerman) [reply src/libopenrave-core/jsonparser/jsonreader.cpp:160]: Removed remotescheme

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 8 above)
Comment (JosephBerman) [reply src/libopenrave-core/jsonparser/jsonreader.cpp]: removed starts with, need to implement other protocols such as ftp://, 
file:// currently works

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 9 above)
Comment (JosephBerman) [reply src/libopenrave-core/jsonparser/jsonreader.cpp]: fixed

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 10 above)
Comment (JosephBerman) [reply src/libopenrave-core/jsonparser/jsonreader.cpp]: removed starts with

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 12 above)
Comment (JosephBerman) [reply src/libopenrave-core/jsonparser/jsonreader.cpp]: Will add

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 13]
```diff
@@ -384,45 +446,51 @@ class JSONReader
         return false;
     }
 
-    boost::shared_ptr<const rapidjson::Document> _GetDocumentFromFilename(const std::string& fullFilename, rapidjson::Document::AllocatorType& alloc)
+    boost::shared_ptr<const rapidjson::Document> _GetDocumentFromFilename(const std::string& referenceUri, rapidjson::Document::AllocatorType& alloc)
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:449]: @JosephBerman Please explain the changes in this function.

Why do you need to add to `_urlsAlreadyStaged`? At this point, if you don't already have it in _rapidJSONDocuments, it is too late.

Why do you treat `msgpack` different from `json`?

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 14]
```diff
@@ -751,16 +828,40 @@ class JSONReader
...
+        }
+
         FOREACH(itConnected, robotInfo._vConnectedBodyInfos) {
             RobotBase::ConnectedBodyInfoPtr& pConnected = *itConnected;
             std::map<RobotBase::ConnectedBodyInfoPtr, std::string>::iterator itProcessedEntry = mapProcessedConnectedBodyUris.find(pConnected);
             if( itProcessedEntry != mapProcessedConnectedBodyUris.end() && itProcessedEntry->second == pConnected->_uri ) {
                 continue;
             }
 
-            if( !_IsExpandableReferenceUri(pConnected->_uri) ) {
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:761]: Why is this check removed?

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 15]
```diff
@@ -901,12 +1001,266 @@ class JSONReader
...
+            }
+        } while(stillRunning != 0);
+
+        return;
+    }
+
+
+
+    /// \brief data type that is passed to curl handler containing information on parsing incoming files
+    struct curlData
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:1200]: We capitalize struct and class names.

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 16]
```diff
@@ -901,12 +1001,266 @@ class JSONReader
...
+    {
+        rapidjson::StringBuffer *buffer = new rapidjson::StringBuffer(); // Internal buffer assigned once curl is set up to download
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:1202]: This is a leak waiting to happen.

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 17]
```diff
@@ -901,12 +1001,266 @@ class JSONReader
...
+        CURL * curl = curl_easy_init(); // Curl handler requires to be initlized
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:1203]: Another leak. Why not `CURL curl`?

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 18]
```diff
@@ -901,12 +1001,266 @@ class JSONReader
...
+        std::string uri;
+    };
+
+    CURLM *_curlMultiHandle; /// < curl multi handler, used to downlod files simultaneously
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:1207]: _curlMultiHandle = nullptr;

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 19]
```diff
@@ -901,12 +1001,266 @@ class JSONReader
...
+    std::vector<boost::shared_ptr<curlData> > _curlDataVector; ///< Holds all curl handlers
+    std::vector<std::string> _urlsAlreadyStaged; ///< Holds the URLs that have already been downloaded or will be
+
+    /// @brief Write back function for libcurl, all data that is retrieves is then pushed into a rapidjson::StringBuffer
+    /// @param data Data recieved from libcurl
+    /// @param size size
+    /// @param nmemb size
+    /// @param userdata Own structure to store incoming data
+    /// @return Returns the total size back to libcurl
+    static size_t _WriteBackDataFromCurl(char *data, size_t size, size_t nmemb, curlData &userdata)
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:1217]: `nmemb` is not a good name.

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 20]
```diff
@@ -362,6 +391,39 @@ class JSONReader
...
+        ParseURI(uri, scheme, path, fragment );
+
+
+        if (IsDownloadingFromRemote()) {
+
+            if (scheme == "file") {
+                return ("file://" + ResolveURI(scheme, path, std::string(), _vOpenRAVESchemeAliases));
+            }
+
+            return _remoteUrl + uri.substr(uri.find_last_of(":/")+1,  uri.find_last_of("#") - uri.find_last_of(":/") -1);
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:410]: Like I mentioned in the last round, `_remoteUrl` should only be used when you `curl_easy_setopt(CURLOPT_URL)`, it should not be used anywhere else in the code.

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 13 above)
Comment (JosephBerman) [reply src/libopenrave-core/jsonparser/jsonreader.cpp:449]: That was left over from when still using ifstream, removing.

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 14 above)
Comment (JosephBerman) [reply src/libopenrave-core/jsonparser/jsonreader.cpp:761]: Readded

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 15 above)
Comment (JosephBerman) [reply src/libopenrave-core/jsonparser/jsonreader.cpp:1200]: `curlData` -> `CurlData`

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 16 above)
Comment (JosephBerman) [reply src/libopenrave-core/jsonparser/jsonreader.cpp:1202]: I have not implemented the rotating string buffers, i however have the .hpp file from a previous implementation

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 21]
```diff
@@ -133,6 +141,424 @@ static std::string CanonicalizeURI(const std::string& suburi, const std::string&
...
+    size_t hashindex = path.find_last_of('#');
+
+    if (hashindex != std::string::npos) {
+        path = path.substr(0, hashindex);
+    }
+    return path;
+}
+
+
+class JSONRemoteHelper
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:158]: @JosephBerman I think you can even pull this out to a separate file in the `jsonreader`, e.g. `jsonreader/jsondownloader.cpp`

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 22]
```diff
@@ -324,11 +771,12 @@ class JSONReader
     {
         std::string scheme, path, fragment;
         ParseURI(uri, scheme, path, fragment);
+
         if (fragment == "") {
             return ExtractFirst(rEnvInfo, ppbody, alloc);
         }
 
-        // find the body by uri fragment
+        // find the body by uri fragmentclear
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:779]: What's this change?

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 23]
```diff
@@ -388,24 +850,25 @@ class JSONReader
     {
         boost::shared_ptr<const rapidjson::Document> doc;
         // TODO: optimize this. for the first time doc is cached, all the expandable object will never get cached, because we are not update document cache after expand any body
+
         if (_rapidJSONDocuments.find(fullFilename) != _rapidJSONDocuments.end()) {
             doc = _rapidJSONDocuments[fullFilename];
         }
         else {
-            boost::shared_ptr<rapidjson::Document> newDoc;
+            boost::shared_ptr<rapidjson::Document> newDoc = boost::shared_ptr<rapidjson::Document>(new rapidjson::Document());;
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:858]: Here you are losing `alloc` again. You don't need these 3 lines of changes. Please revert.

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 24]
```diff
@@ -751,6 +1230,33 @@ class JSONReader
     /// \param[inout] mapProcessedConnectedBodyUris a map of the already processed connected bodies with their URIs. This is to prevent multiple passes from opening the same files
     void _ProcessURIsInRobotBaseInfo(RobotBase::RobotBaseInfo& robotInfo, const rapidjson::Value& rEnvInfo, dReal fUnitScale, rapidjson::Document::AllocatorType& alloc, std::map<RobotBase::ConnectedBodyInfoPtr, std::string>& mapProcessedConnectedBodyUris)
     {
+
+        if(()) {
```
Comment (ziyan) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:1234]: What's this condition?

**Comment by rdiankov** [comment]
@ziyan is this ready for review?

**Comment by ziyan** [comment]
> @ziyan is this ready for review?

I am still working on this.

File: src/libopenrave-core/CMakeLists.txt [hunk 25]
```diff
@@ -16,6 +16,13 @@ if( libpcrecpp_FOUND )
   add_definitions(-DHAVE_PCRECPP)
 endif()
 
+find_package(CURL)
+if ( CURL_FOUND )
+  set(OPENRAVE_CORE_LIBRARIES ${OPENRAVE_CORE_LIBRARIES} ${CURL_LIBRARIES})
+else()
+  message(FATAL_ERROR "curl library not found")
```
Comment (rdiankov) [inline src/libopenrave-core/CMakeLists.txt:23]: @ziyan let's make it possible to use openrave without CURL. I would use `add_definitions` and check for that in the code

File: src/libopenrave-core/jsonparser/jsondownloader.cpp [hunk 26]
```diff
@@ -0,0 +1,330 @@
...
+JSONDownloader::JSONDownloader(rapidjson::Document::AllocatorType& alloc, std::map<std::string, boost::shared_ptr<const rapidjson::Document> >& rapidJSONDocuments, const std::string& remoteUrl, const std::vector<std::string>& vOpenRAVESchemeAliases, bool downloadRecursively) :
+    _alloc(alloc),
+    _rapidJSONDocuments(rapidJSONDocuments),
+    _remoteUrl(remoteUrl),
+    _vOpenRAVESchemeAliases(vOpenRAVESchemeAliases),
+    _downloadRecursively(downloadRecursively)
+{
+    _curlm = curl_multi_init();
+    if (!_curlm) {
+        throw OPENRAVE_EXCEPTION_FORMAT0("failed to create curl handle", ORE_InvalidArguments);
```
Comment (rdiankov) [inline src/libopenrave-core/jsonparser/jsondownloader.cpp:92]: @ziyan let's start giving more meaningful exception codes. Ie ORE_CurlObjectInvalid

File: src/libopenrave-core/jsonparser/jsondownloader.cpp [hunk 27]
```diff
@@ -0,0 +1,330 @@
...
+            if (it == _mapDownloadContexts.end()) {
+                throw OPENRAVE_EXCEPTION_FORMAT0("curl download finished, but failed to find corresponding context, cannot continue", ORE_InvalidArguments);
+            }
+            JSONDownloadContextPtr pContext = it->second;
+            _mapDownloadContexts.erase(it);
+
+            // check result
+            const CURLcode curlCode = curlMessage->data.result;
+            if (curlCode != CURLE_OK) {
+                throw OPENRAVE_EXCEPTION_FORMAT("failed to download uri \"%s\": %s", pContext->uri%curl_easy_strerror(curlCode), ORE_InvalidURI);
```
Comment (rdiankov) [inline src/libopenrave-core/jsonparser/jsondownloader.cpp:168]: ORE_CurlInvalidResponse

File: src/libopenrave-core/jsonparser/jsondownloader.cpp [hunk 28]
```diff
@@ -0,0 +1,330 @@
...
+            }
+
+            // check http response code
+            int responseCode = 0;
+            const CURLcode getInfoCode = curl_easy_getinfo(pContext->curl, CURLINFO_RESPONSE_CODE, &responseCode);
+            if (getInfoCode != CURLE_OK) {
+                throw OPENRAVE_EXCEPTION_FORMAT("failed to get response status code for uri \"%s\": %s", pContext->uri%curl_easy_strerror(getInfoCode), ORE_InvalidArguments);
```
Comment (rdiankov) [inline src/libopenrave-core/jsonparser/jsondownloader.cpp:175]: `ORE_CurlInvalidResponse`

File: src/libopenrave-core/jsonparser/jsondownloader.cpp [hunk 29]
```diff
@@ -0,0 +1,330 @@
...
+            }
+            if (responseCode != 0 && responseCode != 200) {
+                // file scheme downloads have a zero response code
+                throw OPENRAVE_EXCEPTION_FORMAT("failed to download uri \"%s\", received http %d response", pContext->uri%responseCode, ORE_InvalidURI);
```
Comment (rdiankov) [inline src/libopenrave-core/jsonparser/jsondownloader.cpp:179]: `ORE_CurlInvalidResponse`

File: src/libopenrave-core/jsonparser/jsondownloader.cpp [hunk 30]
```diff
@@ -0,0 +1,330 @@
...
+    }
+    curlCode = curl_easy_setopt(pContext->curl, CURLOPT_WRITEDATA, pContext.get());
+    if (curlCode != CURLE_OK) { 
+        throw OPENRAVE_EXCEPTION_FORMAT("failed to curl_easy_setopt(CURLOPT_WRITEDATA) for uri \"%s\": %s", canonicalUri%curl_easy_strerror(curlCode), ORE_InvalidArguments);
+    }
+
+    // add handle to curl multi
+    const CURLMcode addHandleCode = curl_multi_add_handle(_curlm, pContext->curl);
+    if (!!addHandleCode) {
+        throw OPENRAVE_EXCEPTION_FORMAT("failed to download uri \"%s\", curl_multi_add_handle() failed with code %d", canonicalUri%(int)addHandleCode, ORE_InvalidArguments);
```
Comment (rdiankov) [inline src/libopenrave-core/jsonparser/jsondownloader.cpp:279]: possibly ORE_CurlInvalidHandle

File: src/libopenrave-core/jsonparser/jsondownloader.cpp (hunk 27 above)
Comment (ziyan) [reply src/libopenrave-core/jsonparser/jsondownloader.cpp:168]: Yes, because the handle in curlMessage was created by curl_easy_init

File: src/libopenrave-core/jsonparser/jsondownloader.cpp (hunk 28 above)
Comment (ziyan) [reply src/libopenrave-core/jsonparser/jsondownloader.cpp:175]: For this case, I think `CurlHandleInvalid` is more appropriate, since this error should not happen unless `pContext->curl` is somehow invalid.

File: src/libopenrave-core/jsonparser/jsondownloader.cpp [hunk 31]
```diff
@@ -0,0 +1,330 @@
...
+    JSONDownloadContextPtr pContext = boost::make_shared<JSONDownloadContext>();
+    pContext->uri = canonicalUri;
+    pContext->pDoc = pDoc;
+    pContext->startTimestampUS = utils::GetMonotonicTime();
+
+    // set curl options
+    CURLcode curlCode;
+    curlCode = curl_easy_setopt(pContext->curl, CURLOPT_URL, url.c_str());
+    if (curlCode != CURLE_OK) { 
+        throw OPENRAVE_EXCEPTION_FORMAT("failed to curl_easy_setopt(CURLOPT_URL) for uri \"%s\": %s", canonicalUri%curl_easy_strerror(curlCode), ORE_InvalidArguments);
```
[the remark this answers is not in the document]
Comment (ziyan) [reply src/libopenrave-core/jsonparser/jsondownloader.cpp:261]: These should be `ORE_CurlHandleInvalid`, the request has not been sent yet, here we are just setting up some parameters. Most likely it can only fail if there is some memory issues.

File: src/libopenrave-core/CMakeLists.txt (hunk 25 above)
Comment (ziyan) [reply src/libopenrave-core/CMakeLists.txt]: Done.

File: src/libopenrave-core/jsonparser/jsondownloader.cpp (hunk 26 above)
Comment (ziyan) [reply src/libopenrave-core/jsonparser/jsondownloader.cpp]: Added `ORE_CurlInvalidHandle`

File: src/libopenrave-core/jsonparser/jsondownloader.cpp [hunk 32]
```diff
@@ -0,0 +1,330 @@
...
+            const CURLMcode pollCode = curl_multi_poll(_curlm, NULL, 0, 100, NULL);
+            if (pollCode) {
+                throw OPENRAVE_EXCEPTION_FORMAT("failed to download, curl_multi_poll() failed with code %d", (int)pollCode, ORE_InvalidArguments);
+            }
+        }
+
+        // check for timeout
+        const uint64_t currentTimestampUS = utils::GetMonotonicTime();
+        if (currentTimestampUS > startTimestampUS + timeoutUS) {
+            throw OPENRAVE_EXCEPTION_FORMAT("timed out waiting for download to finish, timeout is %d[us]", timeoutUS, ORE_Timeout);
```
[the remark this answers is not in the document]
Comment (ziyan) [reply src/libopenrave-core/jsonparser/jsondownloader.cpp]: I added `ORE_CurlTimeout`, maybe there is upload in the future?

But I think `ORE_Timeout` suffice here?

File: src/libopenrave-core/jsonparser/jsondownloader.cpp [hunk 33]
```diff
@@ -0,0 +1,330 @@
...
+            }
+
+            // remove handle
+            const CURLMcode removeHandleCode = curl_multi_remove_handle(_curlm, curlMessage->easy_handle);
+            if (removeHandleCode) {
+                throw OPENRAVE_EXCEPTION_FORMAT("failed to download, curl_multi_remove_handle() failed with code %d", (int)removeHandleCode, ORE_InvalidArguments);
+            }
+            const std::map<CURL*, JSONDownloadContextPtr>::iterator it = _mapDownloadContexts.find(curlMessage->easy_handle);
+            if (it == _mapDownloadContexts.end()) {
+                throw OPENRAVE_EXCEPTION_FORMAT0("curl download finished, but failed to find corresponding context, cannot continue", ORE_InvalidArguments);
```
[the remark this answers is not in the document]
Comment (ziyan) [reply src/libopenrave-core/jsonparser/jsondownloader.cpp]: I used `ORE_CurlInvalidHandle` because this is not a normal case, this case should never happen, or there is a bug in this c++ file.

File: src/libopenrave-core/jsonparser/jsondownloader.cpp (hunk 29 above)
Comment (ziyan) [reply src/libopenrave-core/jsonparser/jsondownloader.cpp]: Added `ORE_CurlInvalidResponse`

File: src/libopenrave-core/jsonparser/jsondownloader.cpp (hunk 30 above)
Comment (ziyan) [reply src/libopenrave-core/jsonparser/jsondownloader.cpp]: Added `ORE_CurlInvalidResponse`

**Comment by ziyan** [comment]
@rdiankov test pipeline #426715

File: src/libopenrave-core/jsonparser/jsondownloader.cpp [hunk 34]
```diff
@@ -0,0 +1,353 @@
...
+        for (;;) {
+            int numMessagesInQueue = 0;
+            const CURLMsg *curlMessage = curl_multi_info_read(_curlm, &numMessagesInQueue);
+            RAVELOG_VERBOSE_FORMAT("curl_multi_info_read(): numMessagesInQueue = %d", numMessagesInQueue);
+            if (!curlMessage) {
+                break;
+            }
+            RAVELOG_VERBOSE_FORMAT("curl_multi_info_read(): curlMessage->msg = %d", (int)curlMessage->msg);
+            if (curlMessage->msg != CURLMSG_DONE) {
+                continue;
```
Comment (rdiankov) [inline src/libopenrave-core/jsonparser/jsondownloader.cpp:154]: @ziyan wouldn't busy waiting take up CPU resources? isn't sleep better?
